<launch>
  <!--
       Required arguments:
          HAND:=right|left
       Suggested arguments:
          NUM:=0|1|...
          ZEROS:=/path/to/zeros_file.yaml
          CONTROLLER:=grasp|pd|velsat|torque|sim
          RESPAWN:=true|false   Respawn controller if it dies.
          KEYBOARD:=true|false  (default is true)
          AUTO_CAN:=true|false  (if true, ignores CAN_DEVICE argument and finds the can device automagically).
          CAN_DEVICE:=/dev/pcanusb1 | /dev/pcanusbNNN  (ls -l /dev/pcan* to see open CAN devices)
          VISUALIZE:=true|false  (Launch rviz)
          JSP_GUI:=true|false  (If true, start a joint_state_publisher for commanded joint angles).
          GAZEBO:=true|false  (NOTE: Gazebo support is highly experimental at this point)

       This script launches the following nodes:
         - allegro hand controller (different controllers exist)
         - keyboard controller
         - state publisher (for TF information)

       NOTE: If you specify the can device manually (CAN_DEVICE:=/dev/pcanusbN),
             make sure you *also* specify AUTO_CAN:=false.

  -->


  <!-- The inclusion of which_hand in the zero.yaml file has been deprecated.
       Which hand (left/right) must now be specified as an argument when launching the Allegro Hand
       as an input for both the robot_description and for the grasping library controllers. -->
  <arg name="HAND" default="right"/>
  <arg name="NUM" default="0"/>
  <arg name="LOG" default="false"/>

  <!-- ls -l /dev/pcan* to see your open CAN ports. Auto means find it
       automatically, and ignores the CAN_DEVICE argument. -->
  <arg name="AUTO_CAN" default="true" />
  <arg name="CAN_DEVICE" default="/dev/pcanusb1" />

  <arg name="PARAMS_DIR" default="$(find allegro_hand_parameters)" />
  <arg name="KEYBOARD" default="true" />

  <!-- yaml param files for your hand can be found in parameters/zero_files/ -->
  <arg name="ZEROS" default="$(arg PARAMS_DIR)/zero.yaml"/>

  <!--If true, respawn the controller if it dies. -->
  <arg name="RESPAWN" default="false"/>

  <!-- Load the robot description directly from the xacro file. (NOTE: store it
       in two parameter names.) -->
  <param name="robot_description"
         command="$(find xacro)/xacro
                  $(find allegro_hand_description)/urdf/allegro_hand_r.xacro"/>

  <!-- Allegro Hand controller and communication node. -->
  <node name="allegro_hand_hardware_interface"
        pkg="allegro_hand_ros_control"
        type="allegro_hand_hardware_interface"
        output="screen"
        clear_params="true"
        respawn="$(arg RESPAWN)"
        respawn_delay="2" >

    <!--parameters are within the scope of the hand node so that multiple hands can be run at the same time -->
    <rosparam file="$(arg ZEROS)" command="load" />

    <rosparam file="$(arg PARAMS_DIR)/gains_pd.yaml" command="load" />
    <rosparam file="$(arg PARAMS_DIR)/initial_position.yaml" command="load" />

    <!-- Set the CAN channel automatically (using detect_pcan.py) if the
         AUTO_CAN parameter is true, otherwise use the CAN_DEVICE
         argument. NOTE: To manually set the can device, you must *also* set
         AUTO_CAN:=false. -->
    <param name="/comm/CAN_CH" value="$(arg CAN_DEVICE)"
           unless="$(arg AUTO_CAN)" />
    <param name="/comm/CAN_CH"
           command="$(find allegro_hand_description)/scripts/detect_pcan.py"
           if="$(arg AUTO_CAN)" />

    <param name="/hand_info/which_hand" value="$(arg HAND)" /> <!-- See HAND arg above -->
  </node>

  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />

  <rosparam file="$(find allegro_hand_ros_control)/config/joint_state_controller.yaml" command="load"/>
  <node name="joint_state_controller_spawner" pkg="controller_manager" type="controller_manager" args="spawn joint_state_controller" respawn="false" output="screen"/>

  <rosparam file="$(find allegro_hand_ros_control)/config/allegro_hand_control.yaml" command="load"/>
  <node name="allegro_hand_controller_spawner" pkg="controller_manager" type="controller_manager" args="spawn allegro_hand_controller" respawn="false" output="screen"/>

</launch>
